src/learning/WeightUpdater.cpp

```diff
@@ -0,0 +1,112 @@
+#include "../../include/learning/WeightUpdater.h"
+#include "../../include/learning/LinearSolver.h"
+
+#include <sstream>
+
+WeightUpdater::WeightUpdater(AlternativesPerformance &ap, Config &conf)
+    : ap(ap), conf(conf), solver(ap, conf) {}
+
+WeightUpdater::WeightUpdater(const WeightUpdater &wu)
+    : solver(wu.solver), ap(wu.ap), conf(wu.conf) {}
+
+WeightUpdater::~WeightUpdater() {}
+
+void WeightUpdater::updateWeightsAndLambda(MRSortModel &model) {
+
+  if (!this->modelCheck(model)) {
+    throw std::invalid_argument("Model's profile doesn't suite the alternative "
+                                "performance of this WeightUpdater");
+  }
+
+  auto matrix_x = this->computeXMatrix(model);
+  auto matrix_y = this->computeYMatrix(model);
+  std::pair<float, std::vector<float>> res = solver.solve(matrix_x, matrix_y);
+
+  std::ostringstream ss;
+  ss << "Linear problem results - ";
+
+  // update lambda
+  model.lambda = res.first;
+  ss << "Lambda: " << res.first;
+  // update all weights in order
+  model.criteria.setWeights(res.second);
+  for (int i = 0; i < res.second.size(); i++) {
+    ss << " - w" << i << ": " << model.criteria[i].getWeight();
+  }
+  conf.logger->info(ss.str());
+}
+
+std::vector<std::vector<std::vector<bool>>>
+WeightUpdater::computeXMatrix(MRSortModel &model) {
+
+  std::vector<std::vector<std::vector<bool>>> x_matrix;
+  auto profs_pt = model.profiles.getPerformanceTable();
+  auto alts_pt = ap.getPerformanceTable();
+  auto alts_assign = ap.getAlternativesAssignments();
+  // As the profiles are stored in descending order (h = 0 is the highest
+  // profile, which is not the case in the theorical formulation), going in
+  // reverse.
+  for (int h = profs_pt.size() - 2; h >= 0; h--) {
+    std::vector<std::vector<bool>> x_h;
```

Comment:
No for the folowing profile
```
    cat0  cat1
b1  0.8  0.6
b0 0.4   0.3
```

in the way we defined the Profiles performance table it would be:
```
prof = [[0.8,0.6],[0.4,0.3]]
```
thus b0 is actually prof[1] and not prof[0].

That's why we are going in reverse if we compare with the thesis